chain/transaction_test.go

```diff
@@ -74,6 +77,73 @@ func unmarshalAction2(p *codec.Packer) (chain.Action, error) {
 	return &action, err
 }
 
+func TestJSONMarshalUnmarshal(t *testing.T) {
+	require := require.New(t)
+
+	txdata := chain.TransactionData{
+		Base: &chain.Base{
+			Timestamp: 1724315246000,
+			ChainID:   [32]byte{1, 2, 3, 4, 5, 6, 7},
+			MaxFee:    1234567,
+		},
+		Actions: []chain.Action{
+			&mockTransferAction{
+				To:    codec.Address{1, 2, 3, 4},
+				Value: 4,
+				Memo:  []byte("hello"),
+			},
+			&mockTransferAction{
+				To:    codec.Address{4, 5, 6, 7},
+				Value: 123,
+				Memo:  []byte("world"),
+			},
+			&action2{
+				A: 2,
+				B: 4,
+			},
+		},
+	}
+	priv, err := ed25519.GeneratePrivateKey()
+	require.NoError(err)
+	factory := auth.NewED25519Factory(priv)
+
+	actionCodec := codec.NewTypeParser[chain.Action]()
+	authCodec := codec.NewTypeParser[chain.Auth]()
+
+	err = actionCodec.Register(&mockTransferAction{}, unmarshalTransfer)
+	require.NoError(err)
+	err = actionCodec.Register(&action2{}, unmarshalAction2)
+	require.NoError(err)
+	err = authCodec.Register(&auth.ED25519{}, auth.UnmarshalED25519)
+	require.NoError(err)
+
+	signedTx, err := txdata.Sign(factory)
+	require.NoError(err)
+
+	b, err := json.Marshal(signedTx)
+	require.NoError(err)
+
+	parser := chaintest.NewParser(nil, actionCodec, authCodec, nil)
+
+	var txout chain.Transaction
+	err = txout.UnmarshalJSON(b, parser)
+	require.NoError(err)
+	// cannot check direct Equal between signedTx and txout
+	// because of the Auth codec.Address that will be different after the unmarshaling.
```

Comment:
Why is auth different after unmarshalling?

Reply:
private field `addr` is not Marshalled, looking at the `address()` function is shows:
```
func (d *ED25519) address() codec.Address {
	if d.addr == codec.EmptyAddress {
		d.addr = NewED25519Address(d.Signer)
	}
	return d.addr
}
```
I see the same logic for BLS, ED25519, SECP256R1.